cpp/src/arrow/ipc/json-internal.cc

```diff
@@ -1094,6 +1108,32 @@ class ArrayReader {
   }
 
   template <typename T>
+  typename std::enable_if<std::is_base_of<DecimalType, T>::value, Status>::type Visit(
+      const T& type) {
+    typename TypeTraits<T>::BuilderType builder(type_, pool_);
+
+    const auto& json_data = obj_->FindMember("DATA");
+    RETURN_NOT_ARRAY("DATA", json_data, *obj_);
+
+    const auto& json_data_arr = json_data->value.GetArray();
+
+    DCHECK_EQ(static_cast<int32_t>(json_data_arr.Size()), length_);
+
+    for (int i = 0; i < length_; ++i) {
+      if (!is_valid_[i]) {
+        RETURN_NOT_OK(builder.AppendNull());
+        continue;
+      }
+
+      const rj::Value& val = json_data_arr[i];
+      DCHECK(val.IsString()) << "JSON value is not a string";
+      const Decimal128 value(val.GetString());
```

Comment:
Can this fail?

Reply:
I'll use the `Decimal::FromString` static method here since that'll return `Status`.